santoku/utils/configuration_manager.py

```diff
@@ -0,0 +1,264 @@
+    formed by settings containing the same keys but different values. The purpose of this class is
+    to handle the definition and retrieving of all the configuartion parameters so that the user can
+    control the execution context of the code.
+
+    """
+
+    def __init__(
+        self,
+        predefined_configurations: List[Dict[str, Any]] = None,
+        default_configuration: str = None,
+    ):
+        """
+        Initialize internal structures that store the configurations.
+
+        A list of `predefined_configurations` can be passed to initialze the configuration manager,
+        together with the name of the `default_configuration`, if it's not given, he first
+        configuration is set by default.
+
+        Parameters
+        ----------
+        predefined_configurations : List[Dict[str, Any]], optional
+            Collection of configurations with different combination of settings.
+        default_configuration: str, optional
+            Name of the configuration in the `predefined_configurations` that will be set as the
+            default one.
+
+        Raises
+        ------
+        UndefinedConfiguration
+            If a `default_configuration` that is not in the `predefined_configurations` is given.
+
+        InvalidConfiguration
+            If any configuration in the `predefined_configurations` does not contain the "NAME" or
+            "CONFIG" argument.
+
+        See also
+        --------
+        apply_configuration: this method is called to set the default configuration.
+
+        Notes
+        -----
+            The following structure is assumed for the list of predefined configurations:
+            [
+                {
+                    "NAME": <name_of_the_configuration>,
+                    "CONFIG": {
+                        <key>: <value>,
+                        ...
+                    }
+                },
+                ...
+            ]
+
+        """
+
+        self._settings: Dict[str, Any] = {}
+        self._configurations: Dict[str, Dict[str, Any]] = {}
+        self._current_configuration: str = ""
+
+        if predefined_configurations:
+
+            for configuration in predefined_configurations:
+
+                if "NAME" not in configuration and "CONFIG" not in configuration:
+                    raise InvalidConfiguration(
+                        "'NAME' and 'CONFIG' arguments must be contained in a configuration."
+                    )
+
+                self.define_configuration(
+                    name=configuration["NAME"], configuration=configuration["CONFIG"]
+                )
+
+            if default_configuration:
+                self.apply_configuration(name=default_configuration)
+            else:
+                # Set the first configuratoin as default.
+                self.apply_configuration(name=next(iter(self._configurations)))
+        elif default_configuration:
```

Comment:
@HENRYQIULO can you explain this case?

Reply:
As both parameters were optional , this was just to prevent a user giving a default configuration without giving a predefined configuration, I think we can just ignore this weird case.